LFS-602: Pagination for Forms

acrowthe opened the pull request:
Implement stepper and pagination
- Next button changes text to save/saving based on page state
- Both back and next buttons save on click and are disabled when saving
- Sections with the paginate property to true will be detected and displayed as a page

Testing should be performed with the [LFS-602-test](https://github.com/ccmbioinfo/lfs/tree/LFS-602-test) branch
- The test branch adds pagination to the LFS demographic and cardiac 0-5NDVariables forms for testing purposes
- This branch also makes the 2nd and 4th page of the demographics forms conditional: Page 2 is shown when Page 1 `Gender` = `Female`, page 4 is shown when Page 3 `Vital status` = `alive and NED`

Missing features:
- Re-import cardiac forms with pagination (Out of scope, future PR)
- Pagination option in form editor (Out of scope, future PR)
- Handle forms with too many pages

acrowthe committed:
LFS-602: Pagination for Forms
Implement stepper and pagination
- Next button changes text to save/saving based on page state
- Both back and next buttons save on click and are disabled when saving
- Sections with the paginate property to true will be detected and displayed as a page

acrowthe committed:
LFS-602: Pagination for Forms
Convert stepper to MobileStepper prgoress bar
- Handles large numbers of pages and small screens better

sdumitriu reviewed (commented):
I think there should only be a `paginate` setting per questionnaire, not for each section. Does it make sense to have only 3 out of 7 sections paginated?

Conditional sections aren't handled very well. For example, adding a condition on "gender = F" to the "Clinical information" section, will lead to an empty page. I'm not sure what's the best way to handle this, though. And things may get even more complicated with repeated sections. The best thing I can come up with is to ditch section names and just use a progress bar that assigns a fixed chunk to each section, and which will just jump over disabled sections, and subdivide the chunk in case of repetitions. As a plus, it also neatly solves the case of too many sections.

JS error if multiple sections don't have a name (or have the same name).

I'm a bit irked by the fact that sections build the `pageList` and decide if they should display themselves or not. I feel like this should be a separate responsibility handled separately, but I'm not sure my concern is valid or not yet.

sdumitriu commented:
(review was done before the last change which introduced the mobile stepper)

acrowthe commented:
The change to the mobile stepper likely only changed the UI: the pagination still uses names to track pages, just doesn't display them anymore

sashaandjic commented:
NEXT and BACK buttons work, however they should load the next/previous page and position to the top field (not at the random position as it appearzs now).

sashaandjic commented:
Also, I am kind of used to seeing SAVE button all the time. SAVE button now is visible on the last page only. If we have auto save this should not be an issue.

sdumitriu commented:
> If we have auto save this should not be an issue.

Back and Next act as Save buttons as well. Every time you change the page, the form is saved. Is that autosave enough?

But the Save button at the end is a bit misleading. Saving happens at every step, and as it is, it serves little purpose. Maybe have a "Save and Close" instead, which would go back to the dashboard?

sashaandjic commented:
> > If we have auto save this should not be an issue.
> 
> Back and Next act as Save buttons as well. Every time you change the page, the form is saved. Is that autosave enough?
> 
> But the Save button at the end is a bit misleading. Saving happens at every step, and as it is, it serves little purpose. Maybe have a "Save and Close" instead, which would go back to the dashboard?

Next/Back buttons doing save is good. I do not like having SAVE option available only on the last page. 

It bothers me is that we do not prompt user to save the unsaved data entries. For example I may go to the demographic form and change some fields on a page, then click on the dashboard. My change will not be saved and as a user I won't know. But this is out of scope of this PR.

marta- commented:
> But the Save button at the end is a bit misleading. Saving happens at every step, and as it is, it serves little purpose. Maybe have a "Save and Close" instead, which would go back to the dashboard?

I like the idea of a button that takes you back (not to the dashboard, but to where you came from, could be the patient chart, the forms homepage, etc).

IMO saving should be done automatically when leaving the page anyway (maybe as part of a separate task). So with that in mind and following the "rule" that button labels should communicate their purpose in as few words as possible, I would suggest "Done" instead of "Save and close" for the final save button. Thoughts?

sashaandjic commented:
Done sounds better. But now I am thinking if we need the button at all?

sashaandjic commented:
> IMO saving should be done automatically when leaving the page anyway (maybe as part of a separate task).
https://phenotips.atlassian.net/browse/LFS-635

marta- commented:
> So with that in mind and following the "rule" that button labels should communicate their purpose in as few words as possible, I would suggest "Done" instead of "Save and close" for the final save button. Thoughts?

There's another best practice regarding button labels, ideally they should be actions. So "Finish" is another option.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -220,6 +223,17 @@ const questionnaireStyle = theme => ({
     },
+    formFooter: {
+        position: "sticky",
+        bottom: 0,
```
I don't like how the form can be seen in the background under the sticky progress bar.
![captured (7)](https://user-images.githubusercontent.com/651980/96658126-c9721200-1311-11eb-9580-315e984aaf5a.gif)
What about:
```suggestion
        bottom: theme.spacing(-2),
```
to compensate for the bottom padding?

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormPagination.jsx:
```diff
@@ -0,0 +1,107 @@
+      lastSaveStatus === false ? 'Save failed, log in and try again?' :
+      activePage < pages.length -1 ? "Next" :
+      lastSaveStatus && savedLastPage ? 'Saved' :
+      'Save'}
```
Save -> Finish, clicking on it goes back one step in history

acrowthe replied:
To be performed in LFS-641

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormPagination.jsx:
```diff
@@ -0,0 +1,107 @@
+/**
+ * Component that displays an page of a Form.
+ */
+function FormPagination (props) {
```
It confuses me that when you open the first page the progress bar is completely empty, and when you open the last page it's completely full.

acrowthe replied:
Changed so that page 1 has a partially complete bar and the last page has a partially empty bar.

marta- reviewed (changes requested):
Not done reviewing, just posting the first comments

acrowthe committed:
LFS-602: Pagination for Forms
Move pagination logic from sections to forms
- Paginate on a form by form basis
- Do not switch pages to a conditionally hidden page

acrowthe commented:
Known issue: The UI does not properly update the text on the next/save button if the current page is no longer the final page or becomes the final page due to changing conditionals

Reproduction: Using the LFS-602-test branch, navigate to the testing page. The next/save button should show "save", since the final page is hidden by a conditional.
Enter `Vital status` = `alive and NED`, which makes the final page visible.
Expected Result: "save" button changes text to "next" and behaves like a next button
Actual result: Text does not update, but behavior changes to that of a next button.

I encountered issues trying to resolves this since the conditional sections and paginate buttons are in different nested locations

sashaandjic commented:
Is this expected (the progress bar not completely full) on the last page before and after hitting SAVE button.
![image](https://user-images.githubusercontent.com/23243244/97493856-f343be00-193b-11eb-9f9d-f61b0410a549.png)

![image](https://user-images.githubusercontent.com/23243244/97493814-e626cf00-193b-11eb-91c9-1f57959e4315.png)

acrowthe commented:
@sashaandjic The progress bar was changed to be partially empty on the last page per @marta- 's review:
> It confuses me that when you open the first page the progress bar is completely empty, and when you open the last page it's completely full.

marta- commented:
> Is this expected (the progress bar not completely full) on the last page before and after hitting SAVE button

TBH I'm not sure how to best handle this without a "current page" marker (a segment of a different color between the "done" and "to do" segments on the progress bar. Open to feedback and suggestions.

sashaandjic commented:
OK. I have no other complains.

sashaandjic reviewed (approved):
tested

marta- commented:
My suggestion to shift the progress bar didn't improve anything, on the contrary. It will need to be readjusted.

Please see below a progress bar from a mobile app with 1 question per screen.

Question 1/17:

![Screenshot_20201105-140352](https://user-images.githubusercontent.com/651980/98287152-f7429200-1f72-11eb-83b8-249510c93f84.jpg)

Question 2/17:

![Screenshot_20201105-140947](https://user-images.githubusercontent.com/651980/98287203-088b9e80-1f73-11eb-9faf-a83798a2401a.jpg)

Question 17/17:

![Screenshot_20201105-140812](https://user-images.githubusercontent.com/651980/98287274-20632280-1f73-11eb-9abe-56edfa89d035.jpg)

Basically it starts with a larger chuck, and progresses such that at the last page it really looks like you're almost done.

Wdyt @sdumitriu @sashaandjic @acrowthe ?

marta- reviewed (dismissed):
Strange side effect:

![image](https://user-images.githubusercontent.com/651980/98289407-24dd0a80-1f76-11eb-906f-db815540c281.png)
![image](https://user-images.githubusercontent.com/651980/98290076-24913f00-1f77-11eb-8b4f-09898e4f3710.png)


On the patient chart, the summary of a paginated form where I entered only 1 field displays that field repeated 4 times. If I enter 2 fields the sequence is repeated twice (Note that normally we display the first 4 fields, fewer if we don't have 4 filled in). Moreover, the form is not labeled as **incomplete**, even though the mandatory **gender** field is left blank. This does not happen on `dev`.

acrowthe commented:
It is possible to overlap two steppers to add a "current page" section. It involves making multiple sections transparent, and then offsetting one of the stepper by a page. Colors are obviously adjustable.
![image](https://user-images.githubusercontent.com/20056751/98296858-8d7db480-1f81-11eb-9e23-ec37a3ce3143.png)

I'd then be able to make the first and last pages look like this:
![image](https://user-images.githubusercontent.com/20056751/98297333-4217d600-1f82-11eb-8e0f-d21bd15aa313.png)
![image](https://user-images.githubusercontent.com/20056751/98297474-768b9200-1f82-11eb-9572-5ab9bbb138ac.png)

Marta's suggestion with the larger page 1 bar and a slightly empty bar for the last page would also be possible and easier if desired.

marta- commented:
> It is possible to overlap two steppers to add a "current page" section. It involves making multiple sections transparent, and then offsetting one of the stepper by a page. Colors are obviously adjustable.

I really like what I'm seeing here. If it's wrapped in our own progressbar component which gets some meaningful parameters and colors as input and does the appropriate offsetting, it won't even look that hackish.

Note on colors: by default I suggest blue (primary) for "done already", and either orange (warning) or pink (secondary) for "current".

sashaandjic commented:
@marta- @acrowthe I can confirm the weird anomaly where if only one field is entered, the chart will show 4 copies of it. For 2 fields 2 copies (2 visible copies). For 3 fields - 2 (visible) copies 
![image](https://user-images.githubusercontent.com/23243244/98300709-a8ebbe00-1f87-11eb-9fda-5be758d886b2.png)

acrowthe commented:
> I can confirm the weird anomaly where if only one field is entered, the chart will show 4 copies of it. For 2 fields 2 copies (2 visible copies). For 3 fields - 2 (visible) copies

I have found the source of this issue, and have a fix in progress. The subject page isn't handling questionnaires with multiple sections properly

marta- commented:
> > I can confirm the weird anomaly where if only one field is entered, the chart will show 4 copies of it. For 2 fields 2 copies (2 visible copies). For 3 fields - 2 (visible) copies
> 
> I have found the source of this issue, and have a fix in progress. The subject page isn't handling questionnaires with multiple sections properly

Sounds like it's not directly related to pagination and could happen to any sectioned questionnaire. In that case it should be addressed as a separate task.

acrowthe commented:
> If it's wrapped in our own progressbar component ... it won't even look that hackish.

Don't look too far into the component: It is going to have to have 2 back buttons, 2 next buttons and quite hacky CSS to get alignment right and the correct elements hidden

acrowthe commented:
> Sounds like it's not directly caused by pagination and could happen to any sectioned questionnaire. In that case it should be addressed as a separate task.

I've got it half fixed already, and suspect the change will only be a couple lines long. If you'd like I can push that change into a new branch and issue when I am done with it.

marta- commented:
> I've got it half fixed already, and suspect the change will only be a couple lines long. If you'd like I can push that change into a new branch and issue when I am done with it.

Ok, the subject chart issue can go in this PR so it can be tested more easily. However, if the incomplete form issue is deeper it might need a separate task.

acrowthe committed:
LFS-602: Pagination for Forms
Improve subject page handling of forms with sections

acrowthe commented:
Subject section fix is pushed. Ended up being -8+7 in Subject.jsx. If that's a bit large I could still move it into it's own PR.

Incomplete form issue and progress bar improvement are still to come

acrowthe committed:
LFS-602: Pagination for Forms
Insert "current page" section of the progress bar

acrowthe commented:
Current page implemented.
First page of a form:
![image](https://user-images.githubusercontent.com/20056751/98575493-45afa380-2287-11eb-93d3-a582a8d17a9f.png)

Last page of a form:
![image](https://user-images.githubusercontent.com/20056751/98575554-54965600-2287-11eb-8fcd-c3f8be1b0249.png)

acrowthe commented:
> Moreover, the form is not labeled as incomplete, even though the mandatory gender field is left blank.

Looks like the sections are getting marked as incomplete, but the root form is not checking the child sections. This would be an issue with all sectioned forms. Looking into a solution which should be fairly small.

sashaandjic commented:
Can we make the line full blue once SAVE turns into SAVED? Does it make sense?
![image](https://user-images.githubusercontent.com/23243244/98589683-a517ae80-229b-11eb-8f9f-4a34674ee681.png)

sashaandjic commented:
Subject chart issue problem is fixed. 
I was not able to reproduce it in the dev branch. Should I created a separate task for it anyway?

acrowthe commented:
The issue with answers being repeated should probably be logged, as I believe it exists in dev and just hasn't been encountered due to forms not meeting the requirements to trigger it (multiple sections).

sashaandjic commented:
Here are the 3 issues that we found while testing this branch. First 2 are unrelated to this PR, however it was easy to include the foxes in this PR and test it:
- In some situations the chart page will shows repeated fields https://phenotips.atlassian.net/browse/LFS-696
- Forms with conditional sections are not properly marked as COMPLETE or INCOMPLETE https://phenotips.atlassian.net/browse/LFS-697
- [The UI does not properly update the text on the next/save button if the current page is no longer the final page or becomes the final page due to changing conditionals](https://phenotips.atlassian.net/browse/LFS-695)

acrowthe committed:
LFS-602: Pagination for Forms
Fix area above stepper thatt cannot be clicked
Fill bar completely when saved

acrowthe commented:
Adjusted the bar to be complete when the right button says "saved". Will revert to mostly complete if something is changed and the form is no longer completely saved.

![image](https://user-images.githubusercontent.com/20056751/98696771-33df0680-2342-11eb-9bcb-dacd3962ccaf.png)
![image](https://user-images.githubusercontent.com/20056751/98696803-3b061480-2342-11eb-96b2-4d97548755dd.png)

Also fixed an issue where the area above the stepper was not interactive.
